### ISSUE TITLE
Bug: You can't search a room with names like "x.y" reported by  @Puneet-here 

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#12026]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Create a workspace > Rename it to "x.y"
2. Search the room by typing "x.y"

## Expected Result:
You should see the room

## Actual Result:
You won't get any results

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** v1.2.18-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/196831239-02bffcff-2d10-4c5b-a9cc-e63dfdc7cc41.mov

https://user-images.githubusercontent.com/43996225/196831202-373ab346-8920-41e5-a55e-484c10fe6ef7.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @Puneet-here 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1666216269896569

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by laurenreidexpensify** [comment]
@cristipaval looks like a solid external to me.

**Comment by cristipaval** [comment]
Hey @laurenreidexpensify, yes, I had a glance locally on my machine and seems that the search is performed on the client, no api call. We can mark it as external.

**Comment by laurenreidexpensify** [comment]
https://www.upwork.com/jobs/~0105f9e9709c5f9cc1

**Comment by roryabraham** [comment]
I'm not sure ... why don't we just remove periods from both the searchValue and the searchWords? Also, why do we just remove periods and replace commas with a space?

**Comment by roryabraham** [comment]
Long-term, I think we need a more generic, performant, and robust text searching algorithm. We can treat that as out-of-scope for now, but I don't want to add a bunch of domain-specific search rule exceptions, like "ignore periods when looking for DMs, but don't ignore them when looking for chat rooms"

**Comment by roryabraham** [comment]
Overall I'm not confident in any of the proposals so far, sorry

**Comment by laurenreidexpensify** [comment]
Price doubled

**Comment by laurenreidexpensify** [comment]
Price now $1000, bumped thread https://expensify.slack.com/archives/C01GTK53T8Q/p1667826951304139?thread_ts=1666216269.896569&cid=C01GTK53T8Q 

@Puneet-here do you have any solutions to propose?

**Comment by s77rt** [comment]
### Proposal
```diff
diff --git a/src/libs/OptionsListUtils.js b/src/libs/OptionsListUtils.js
index d23a3e119..b420e878c 100644
--- a/src/libs/OptionsListUtils.js
+++ b/src/libs/OptionsListUtils.js
@@ -189,7 +189,7 @@ function getSearchText(report, reportName, personalDetailList, isChatRoomOrPolic
     if (!isChatRoomOrPolicyExpenseChat) {
         for (let i = 0; i < personalDetailList.length; i++) {
             const personalDetail = personalDetailList[i];
-            searchTerms = searchTerms.concat([personalDetail.displayName, personalDetail.login.replace(/\./g, '')]);
+            searchTerms = searchTerms.concat([personalDetail.displayName, personalDetail.login, personalDetail.login.replace(/\./g, '')]);
         }
     }
     if (report) {
@@ -389,14 +389,11 @@ function createOption(logins, personalDetails, report, reportActions = {}, {
  * @returns {Boolean}
  */
 function isSearchStringMatch(searchValue, searchText, participantNames = new Set(), isChatRoom = false) {
-    const searchWords = _.map(
-        searchValue
-            .replace(/\./g, '')
-            .replace(/,/g, ' ')
-            .split(' '),
-        word => word.trim(),
-    );
-    return _.every(searchWords, (word) => {
+    const searchWords = _.compact(_.uniq([
+        searchValue,
+        ..._.map(searchValue.replace(/,/g, ' ').split(' '), word => word.trim()),
+    ]));
+    return _.some(searchWords, (word) => {
         const matchRegex = new RegExp(Str.escapeForRegExp(word), 'i');
         const valueToSearch = searchText && searchText.replace(new RegExp(/&nbsp;/g), '');
         return matchRegex.test(valueToSearch) || (!isChatRoom && participantNames.has(word));
```

### Explanation
The main problem with the previous solutions and the existing one is that the original searchValue is being overwritten and completely lost. The better approach is to add more searchValues and keeping the original.

Following my proposal, in the isSearchStringMatch function, searchWords contains the original searchValue and other derived search values. those derived values are used for two reasons: 1. get more relevant search results and 2. multiple search (see below).

As for the #8007 issue, following the same approach, searching for a member will be based on his display name, his login/email and his filtered login/email (the filter is to remove the periods) as seen on getSearchText function

Also, there is another issue, which is the use of `_every` instead of `_some`. The use of _every was introduced by #1523 to fix #1283 However, as the old proposal only returns fully matched results. using _some will return both partially and fully matched results; meaning if you search for "test1 test2" you will get messages with user test1, test2, test1&test2. Besides that the use of _some is required here nevertheless.

### Results:
1. Searching for "x.y" returns "x.y" (fixes this issues)
2. Searching for "abcxyz" or "abc.xyz" returns emails that contains "abcxyz" and/or "abc.xyz" (fixes #8007)
3. Searching for "test1 test2" or "test1, test2" returns results containing "test1", "test2" and results that contains both "test1" and "test2" (related #1283)


https://user-images.githubusercontent.com/16493223/200965219-3736327a-3d17-4dca-82b5-c827b0bbe59f.mp4

**Comment by laurenreidexpensify** [comment]
@rushatgabhane what you think?

**Comment by s77rt** [comment]
that's correct, this is more a matter of design whether this should be the expected behaviour or not.
Let me know if I should update my proposal given the expected behaviour

**Comment by s77rt** [comment]
@rushatgabhane I can't reproduce the bug. Can you give more details?


https://user-images.githubusercontent.com/16493223/201962311-32cff3b4-ca81-4022-a654-5b1f4e8ecf1e.mp4

**Comment by s77rt** [comment]
@rushatgabhane Still no bug

https://user-images.githubusercontent.com/16493223/201983792-a4d7cbc7-2bc8-4e4d-bf9f-4044be90a31e.mp4

Can you try with the latest version (main branch) then apply the patch

**Comment by roryabraham** [comment]
Overall I think @s77rt's proposal is looking better, but let's make sure we address the problem that @rushatgabhane's seeing

**Comment by laurenreidexpensify** [comment]
Priced increased

**Comment by s77rt** [comment]
@laurenreidexpensify Too early to double I think.
My proposal has not been evaluated yet, waiting for @rushatgabhane update

**Comment by s77rt** [comment]
### Proposal (Updated)
```diff
diff --git a/src/libs/OptionsListUtils.js b/src/libs/OptionsListUtils.js
index d23a3e119..07d600b09 100644
--- a/src/libs/OptionsListUtils.js
+++ b/src/libs/OptionsListUtils.js
@@ -149,7 +149,7 @@ function getParticipantNames(personalDetailList) {
 }
 
 /**
- * A very optimized method to remove unique items from an array.
+ * A very optimized method to remove duplicates from an array.
  * Taken from https://stackoverflow.com/a/9229821/9114791
  *
  * @param {Array} items
@@ -189,7 +189,7 @@ function getSearchText(report, reportName, personalDetailList, isChatRoomOrPolic
     if (!isChatRoomOrPolicyExpenseChat) {
         for (let i = 0; i < personalDetailList.length; i++) {
             const personalDetail = personalDetailList[i];
-            searchTerms = searchTerms.concat([personalDetail.displayName, personalDetail.login.replace(/\./g, '')]);
+            searchTerms = searchTerms.concat([personalDetail.displayName, personalDetail.login, personalDetail.login.replace(/\./g, '')]);
         }
     }
     if (report) {
@@ -389,14 +389,11 @@ function createOption(logins, personalDetails, report, reportActions = {}, {
  * @returns {Boolean}
  */
 function isSearchStringMatch(searchValue, searchText, participantNames = new Set(), isChatRoom = false) {
-    const searchWords = _.map(
-        searchValue
-            .replace(/\./g, '')
-            .replace(/,/g, ' ')
-            .split(' '),
-        word => word.trim(),
-    );
-    return _.every(searchWords, (word) => {
+    const searchWords = _.compact(uniqFast([
+        searchValue,
+        ..._.map(searchValue.replace(/,/g, ' ').split(' '), word => word.trim()),
+    ]));
+    return _.some(searchWords, (word) => {
         const matchRegex = new RegExp(Str.escapeForRegExp(word), 'i');
         const valueToSearch = searchText && searchText.replace(new RegExp(/&nbsp;/g), '');
         return matchRegex.test(valueToSearch) || (!isChatRoom && participantNames.has(word));
```

### Details
Applied @rushatgabhane; Using `uniqFast()` instead of `_.unique()`
Fixed the description (comments) of uniqFast function.

**Comment by laurenreidexpensify** [comment]
bump @roryabraham ^^

**Comment by tc8appdevelo** [comment]
Are you still working on this one? I was thinking about putting in a proposal but then I finally got to the end of the comments and realized it looks like its already about to be done.

**Comment by laurenreidexpensify** [comment]
👋 @roryabraham ^^

**Comment by cead22** [comment]
I'm new to this issue and reviewing proposals, so I welcome any feedback, but I have a few notes and questions
- I don't understand the change in getSearchText, or why we need to add `personalDetail.login,` to `searchTerms` since users logins aren't related to the workspace name or the chat subtitle as far as I know
- Removing the line [`.replace(/\./g, '')`](https://github.com/Expensify/App/blob/00d7c619c/src/libs/OptionsListUtils.js#L389) fixes the issue for me without any other changes but it breaks the tests, and I'm not sure why we're removing period in there, but this looks like a function that is used in many places so there's probably some code using that logic, but since there's no comment, it's hard to say
- Unrelated to this proposal but should this line [`const valueToSearch = searchText && searchText.replace(new RegExp(/&nbsp;/g), '');`](https://github.com/Expensify/App/blob/00d7c619c/src/libs/OptionsListUtils.js#L396) be outside of that loop? 
- The diff below solves the issue and keeps the tests passing, but it changes `_.every` to `_.some` (like @s77rt suggested as well), and I'm not sure if there's a reason why we're using `_.every`

```diff
diff --git a/src/libs/OptionsListUtils.js b/src/libs/OptionsListUtils.js
index faac86e5f..1b2adb1c4 100644
--- a/src/libs/OptionsListUtils.js
+++ b/src/libs/OptionsListUtils.js
@@ -384,14 +384,15 @@ function createOption(logins, personalDetails, report, reportActions = {}, {
  * @returns {Boolean}
  */
 function isSearchStringMatch(searchValue, searchText, participantNames = new Set(), isChatRoom = false) {
-    const searchWords = _.map(
-        searchValue
-            .replace(/\./g, '')
-            .replace(/,/g, ' ')
-            .split(' '),
-        word => word.trim(),
-    );
-    return _.every(searchWords, (word) => {
+    const searchWordsOriginal = searchValue
+        .replace(/,/g, ' ')
+        .split(' ');
+    const searchWordsWithoutPeriods = searchValue
+        .replace(/\./g, '')
+        .replace(/,/g, ' ')
+        .split(' ');
+    const searchWords = searchWordsOriginal.concat(searchWordsWithoutPeriods).map(word => word.trim());
+    return _.some(searchWords, (word) => {
         const matchRegex = new RegExp(Str.escapeForRegExp(word), 'i');
         const valueToSearch = searchText && searchText.replace(new RegExp(/&nbsp;/g), '');
         return matchRegex.test(valueToSearch) || (!isChatRoom && participantNames.has(word));

```

**Comment by laurenreidexpensify** [comment]
@s77rt @rushatgabhane ^^ can you pls address @cead22's questions? Thanks

**Comment by s77rt** [comment]
@cead22 
1. We want to be able to search users too, not just workplaces, chats. We search users by their display name, email and email without dots.
2. This was used to solve this issue https://github.com/Expensify/App/issues/8007 . I have removed it and used instead (1)
3. Indeed, it seems redundant to calculate the same value over and over, I think it should be moved.
4. You are removing the periods for every single case. for performances reasons it's better not to alter the `searchWords` but the `searchTerms`. Don't alter what you use to look for, but alter what you look for. Sorry if this a bit confusing.
> Following my proposal, in the isSearchStringMatch function, searchWords contains the original searchValue and other derived search values. those derived values are used for two reasons: 1. get more relevant search results and 2. multiple search (see below).

**Comment by cead22** [comment]
> 4\. You are removing the periods for every single case. for performances reasons it's better not to alter the `searchWords` but the `searchTerms`. Don't alter what you use to look for, but alter what you look for. Sorry if this a bit confusing.

I don't follow this, can you rephrase?

Also, do you need to update your proposal based on these last few comments?

**Comment by s77rt** [comment]
@cead22 
Search without periods is only needed for emails. That's why I added another value to `getSearchText`.

To explain what I meant, here is an example: search for "hello.test":

- With my proposal:
    1. Check if "hello.test" is included in chat
    2. Check if "hello.test" is included in workspace names
    3. Check if "hello.test" is included in user display name
    4. Check if "hello.test" is included in user email
    5. Check if "hello.test" is included in user email without periods (A)

- With your proposal
    1. Check if "hello.test" is included in chat
    2. Check if "hello.test" is included in workspace names
    3. Check if "hello.test" is included in user display name
    4. Check if "hello.test" is included in user email
    5. Check if "hellotest" is included in chat (B)
    6. Check if "hellotest" is included in workspace names (B)
    7. Check if "hellotest" is included in user display name (B)
    8. Check if "hellotest" is included in user email (B)


The search queries marked with (A) is what I'm adding and the search queries marked with (B) is what you are adding

To further explain what i meant:
**Check if X is included in Y**
I meant to not modify X but instead add another value of Y
As you saw in the above example, modifying X is harmful for the performance and results in unnecessary search queries.

___

No, I don't need to update my proposal, except the part of moving the `valueToSearch` outside the loop which can be handled in the PR.

**Comment by roryabraham** [comment]
Discussed with @cead22, passing this off to him for now

**Comment by cead22** [comment]
@s77rt thanks for the explanation, that makes sense. 

Understanding that, I was wondering why we compute some "variations" in `getSearchText` (ie, `personalDetail.login.replace(/\./g, '')`) and some in `isSearchStringMatch`, to see if we could calculate all the variations in one method, and do the search in the other, but I see `isSearchStringMatch` is used in several places and we'd need to update those, so not following that idea, but I want to check on another one.

I went looking for other usages of `getSearchText` and I only found one usage outside of `OptionsListUtils.js` in `SidebarUtils.js`.

That usage is `result.searchText = OptionsListUtils.getSearchText(...`, and as far I as I can tell, we don't use `result.searchText` anywhere in the component, and don't pass `result` to any other component.

If that's the case
- Can we remove `result.searchText = OptionsListUtils.getSearchText(...)` from `src/libs/SidebarUtils.js`  [here](https://github.com/Expensify/App/blob/0fcebd329/src/libs/SidebarUtils.js#L302)?
- Remove the export of `getSearchText` from `src/libs/OptionsListUtils.js` [here](https://github.com/Expensify/App/blob/0fcebd329/src/libs/OptionsListUtils.js#L805)

Regardless, I won't block on that since that's unrelated to this bug and it's more of an improvement to leave things better than we found them. Given that, @s77rt please go ahead and submit a PR with your proposal 👍

**Comment by cead22** [comment]
@s77rt on second thought, disregard the two bullets above and simply submit your proposal. We can look into updating that separately from this. Thanks!

**Comment by laurenreidexpensify** [comment]
Offers sent to @Puneet-here @s77rt @rushatgabhane in Upwork

**Comment by s77rt** [comment]
@laurenreidexpensify Can you update the offer and add 50% bonus, there is a high chance the PR get merged within 3 days.
[PR](https://github.com/Expensify/App/pull/13186) is ready

**Comment by laurenreidexpensify** [comment]
@s77rt if a bonus is applicable it will be added upon final payment 👍🏽 The offers are sent now for the core job so that we can issue payment as promptly as possible in Upwork 7 days after deploy to production with no regressions.

**Comment by s77rt** [comment]
@laurenreidexpensify PR has been merged. Am I eligible for the 50% bonus? if so please update the offer

**Comment by laurenreidexpensify** [comment]
Payments issued ✔️

Not eligible for bonus because of regression - https://github.com/Expensify/App/issues/13286
